cmd/object-handlers.go

```diff
@@ -435,6 +435,16 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 
 	objInfo := gr.ObjInfo
 
+	// Automatically remove the object/version is an expiry lifecycle rule can be applied
+	if lc, err := globalLifecycleSys.Get(bucket); err == nil {
+		_, ruleApplied := applyLifecycleRules(ctx, *lc, []lifecycle.Action{lifecycle.DeleteAction, lifecycle.DeleteVersionAction}, objectAPI, objInfo, false)
```

Comment:
This implementation will not work @vadmeste you are holding a read lock and then trying to delete with a write lock, you need to expire the object at object layer not at the handler layer. 

It should be done closer to where the lock is held..

Reply:
Or we can do something like an asynchronous style where you do this in a two step process 

-  `applyLifecycleRules() return the Action that was applied but do not Delete yourself, instead
- if the rule was applied return '404' to client and then do `go obj.DeleteObject()` asynchronously.

The advantages are that we will get to the client quickly and not loose on IOPs

Reply:
> This implementation will not work @vadmeste you are holding a read lock and then trying to delete with a write lock,

yeah you are right, I tested with `mc cat` and it works but it turned out `mc cat` issues HEAD request before GET and that removed the object even before calling GetObjectNInfo.

> The advantages are that we will get to the client quickly and not loose on IOPs

Yeah, I thought about this as well, we might use that for listing as well in the future.